### ISSUE TITLE
I'm call it the Phantom Menace

### ISSUE THREAD

**Issue by RyanMullins** [hammerjs/hammer.js#742]
**tl;dr**: Hammer sends a Tap event for each TouchEvent and MouseEvent it intercepts from the browser for every touch a user makes on the screen. Solution is to add [PreventGhostClicks](https://gist.github.com/jtangelder/361052976f044200ea17)-like filtering to the `domHandler()` function that is only active if you are actually using a Touch-enabled device.

**Related Issues**: #727, #736

**Details**: So this all started when I was making changes to Angular Hammer to add an option that [prevents ghost clicks](https://gist.github.com/jtangelder/361052976f044200ea17). And after a really long afternoon of digging, I've finally got an actionable bug report. 

For a long time, we have been having trouble handling Tap events when we clicked on a button. Our app is a big geovisualization project designed for teams of emergency manager to interact with spatial data on big (42" or larger) touch screens. At the top of the app we have a title/menu bar. On the right-hand side of the menu bar we have a button that, when clicked, adds a magnifying-glass-like widget on top of the map. The problem we were having was that when you click on this button, in certain circumstances, two of these magnifying glasses would appear on the screen. So we started looking into this problem. 

Here's what the code for the button looked like when we first started exploring this issue. We have a Glyphicon search icon inside of an anchor, and that anchor has two Angular Attribute Directives that use the `addMagnifier` function to add a magnifier to the UI. 

```html
<a id="toolbarMagnifierButton" 
  hm-tap="addMagnifier()" 
  ng-click="addMagnifier()">
  <span class="glyphicon glyphicon-search" style="font-size:25px;" ></span>
</a>
```

We started by looking at the events that were triggering the `addMagnifier` function. When we threw in some good-old-fashioned `console.log()` statements we found that the `addMagnifier` function was being triggered by both the Tap event and the Click event every time we tapped on the screen. This led me to implement the prevent ghost clicks functionality in Angular Hammer. So after this update our code looked like this.

```html 
<a id="toolbarMagnifierButton" 
  hm-tap="addMagnifier()" 
  hm-recognizer-options='{"preventGhosts":true}'
  ng-click="addMagnifier()">
  <span class="glyphicon glyphicon-search" style="font-size:25px;" ></span>
</a>
```

Even after this fix, we were still getting two calls to the `addMagnifier` function. When we looked at it again we saw that instead of getting two events from different sources (one from Hammer and one from Angular), we were getting two Tap events from Hammer. And so we dug a bit deeper. We started with our own code, just to be absolutely sure it wasn't us (it wasn't), then moved on to Angular Hammer to make sure it wasn't just the prevent ghost clicks failing (it wasn't). Which led us to Hammer. 

We wanted to track events as they made their way through the hammer pipeline. To do this, we added a unique ID to each TouchEvent that came through the `Input.domHandler()` function, since that's what catches all of the DOM events and pushes them towards various Hammer Recognizers, and we thought we would only be getting TouchEvents. We also added a `console.log()` to the beginning of the `Manager.recognize()` to capture the value of `inputData`, since that is what is actually passed to the registered handlers. Now, when we tapped on the screen we could see that we were only assigning a unique ID to a single TouchEvent, but the Manager was still logging two events from the `recognize()` function. Upon further inspection, we found that Hammer is capturing the **TouchEvent** and a **MouseEvent** (specifically a `mouseup`) from the browser, transforming them both into TouchInput events, and sending them both to recognizers.

Here's the code to insert into _hammer.js_ to recreate our experiment.

```JavaScript

// Insert at line 357 of hammer.js

var uniqueIdCounter = 0;

// Insert at line 375 of hammer.js

if (ev instanceof ) {
  ev.uniqueId = ++uniqueIdCounter;
  console.log('INFO: domHandler captured a TouchEvent', ev);
}

// Insert at line 2191 of hammer.js

console.log('INFO: Manager.recognize handling inputData', inputData);
```

And this is the issue that we would like to have solved. The solution is some sort of sneeze-guard that:  

* won't fire two events from a single tap by checking if the TouchEvent and MouseEvent were fired within some threshold (similar to PreventGhostClicks but temporally);
* and is only active if Hammer is connected to a device that is Touch-enabled.

To the best of our ability, we have determined that the `mouseup` event that is eventually passed as the second Tap is firing _after_ the TouchEvent. 

For the time being I've [implemented my own solution to this problem in a fork](https://github.com/RyanMullins/hammer.js). You can view the changes in [this commit](https://github.com/RyanMullins/hammer.js/commit/ccd40ea25b6fa95fc64c21dc0b11abfa345d81f9), but basically I added a filter to only pass TouchEvents through `domHandler()`. The only reason I can implement this fix is because I can guarantee that our app will only ever be run on Touch-enabled devices (yay contract work!), but I think the solution needs to be more robust long-term. I'm going to keep working on this in my fork and see what I can come up with. Please contact me any time to coordinate efforts.

**Comment by RyanMullins** [comment]
There's a chance that this is related to the Angular.js execution context/run loop. If it is then it's not your fault, but it might still be a Hammer.js problem to solve.

I'm going to experiment with some stuff in Angular Hammer tomorrow. I'll get post an update (with a fiddle or a pen or whatever tool the cool kids are using these days) once I do some testing.

**Comment by runspired** [comment]
I don't work on hammer itself, but I'm the author of ember-mobileTouch which uses hammer.  It's probably your run loop or some other eventing, we he a double tap issue / similar issues until we added the click buster.  I also had to take click out of ember's normal eventing, as ember itself was looking for and delegating a click.

**Comment by RyanMullins** [comment]
That answer is why I think this fix may be best implemented in Hammer. It's not actually a Hammer problem, but it's just way easier if Hammer handles it. And PreventGhostClicks always felt like a short term fix anyway.

**Comment by johnpapa** [comment]
+1

**Comment by revolunet** [comment]
@RyanMullins do you use ngTouch simultaneously ? i guess this can cause some headaches.

**Comment by runspired** [comment]
Closing in favor of #808
